imageUpload.js

```diff
@@ -8,83 +8,102 @@ const express = require('express'),
        @param {string} mimeType - e.g. 'image/png' or 'image/jpeg'
-       @param {string} basePath - URL base path, e.g. '/iso_images/'
+       @param {string} basePath - base URL path to the backend, e.g. '/db/'
+       @param {string} categoryPath - URL path to the backend serving the given category of images, e.g. '/iso_images/'
     **/
     let storage = multer.memoryStorage();
     let upload = multer({ storage: storage });
+    let pixelsToBinary = (pixels) => {
+      // assuming pixels are stored as rgba
+      const result = Buffer.allocUnsafe(pixels.data.length / 4);
+      for (let i = 0; i < pixels.data.length; i += 4) {
+        result.writeUInt8(pixels.data[i], i / 4);
+      }
+      return result;
+    };
 
-    pg.schema.createTableIfNotExists(IMG_TABLE_NAME, function (table) {
-      table.text('id');
-      table.text('category');
-      table.binary('data');
-    }).then(() => {
-      app.get(path.join(basePath, ":img_id"),
-        function (req, res) {
-          pg.select(pg.raw('data'))
-            .from(IMG_TABLE_NAME)
-            .where('id', '=', req.params.img_id)
-            .first()
-            .then((row) => {
-              if (row === undefined)
-                throw ({message: `Image with id=${req.params.img_id} does not exist`});
-              let img_buf = row.data;
-              res.type(mimeType);
-              res.end(img_buf, 'binary');
-            })
-            .catch((e) => {
-              logger.error(e.message);
-              res.status(404).send('Not found');
-            });
-        });
+    const imgTableExists = await db.schema.hasTable(IMG_TABLE_NAME);
+    if (!imgTableExists) {
+      await db.schema.createTable(IMG_TABLE_NAME, function (table) {
+        table.text('id').primary();
+        table.text('category');
+        table.binary('data');
+      });
+    }
 
-      app.post(path.join(basePath, 'upload'), upload.single(fieldName),
-        function (req, res, next) {
-          logger.debug(req.file.originalname);
-          let imgID = crypto.randomBytes(16).toString('hex');
-
-          pg.insert({'id': imgID, 'category': fieldName, 'data': req.file.buffer})
-            .into(IMG_TABLE_NAME)
-            .then((m) => {
-              logger.debug(`${m}`);
-              res.status(201).json({ image_id: imgID });
-            })
-            .catch((e) => {
-              logger.error(e.message);
-              res.status(500).send('Failed to store image');
-            });
-        });
+    app.get(path.join(basePath, categoryPath, ":img_id"),
+      async function (req, res) {
+        const row = await db.select(db.raw('data')).from(IMG_TABLE_NAME).where('id', '=', req.params.img_id).first();
+        try {
+          if (row === undefined) {
+            throw ({message: `Image with id=${req.params.img_id} does not exist`});
+          }
+          const imgBuf = row.data;
+          res.type('application/octet-stream');
```

Comment:
@iprotsyuk it feels a bit weird that the post handler accepts 'img/png' but the get handler returns 'application/octet-stream'. Any ideas on that?

Reply:
I certainly see some inconsistency here, but this implementation doesn't really store images. It can set the 'img/png' header for GET responses, but it won't be a valid image.

Reply:
That's the exactly the point. Why not post already binary data then? It will require some minor changes in the engine though.

Reply:
That's the thing, I didn't intend to update the engine because of the image storage optimizations. It can be considered though.